File: algorithms/2_medium/M/02202_maximize_the_topmost_element_after_k_moves/solution.c
inline static int max_arr(const int *restrict nums, const size_t len) {
    int max = INT_MIN;
    for (size_t i = 0; i < len; i += 1) {
        max = MAX(max, nums[i]);
    }
    return max;
}

int maximumTop(const int *restrict nums, const int len, const int k) {
    if (len == 0 || nums == NULL || (len == 1 && (k & 1) == 1)) { return -1; }

    if (len > 1 && k == 1) { return nums[1]; }

    if (k > len) { return max_arr(nums, (size_t)len); }

    int res = -1;
    int max = -1;

    for (int i = 0; i < k; i += 1) {
        res = max;
        max = MAX(max, nums[i]);
    }

    if (k < len) {
        res = MAX(res, nums[k]);
    }

    return res;
}

File: algorithms/2_medium/M/02202_maximize_the_topmost_element_after_k_moves/solution.rs
impl Solution {
    pub fn maximum_top(nums: Vec<i32>, k: i32) -> i32 {
        let k = k as usize;

        if nums.is_empty() || (nums.len() == 1 && (k & 1) == 1) { return -1; }

        if nums.len() > 1 && k == 1 { return nums[1]; }

        if k > nums.len() { return *nums.iter().max().unwrap(); }

        let mut res: i32 = -1;
        let mut max: i32 = -1;

        for i in 0..k {
            res = max;
            max = max.max(nums[i]);
        }

        if k < nums.len() {
            res = res.max(nums[k]);
        }

        res
    }
}
